contract/models/contract_line.py

```diff
@@ -0,0 +1,819 @@
+            new_date = old_date + self.get_relative_delta(
+                line.recurring_rule_type, line.recurring_interval
+            )
+            line.recurring_next_date = new_date
+
+    @api.model
+    def get_relative_delta(self, recurring_rule_type, interval):
+        if recurring_rule_type == 'daily':
+            return relativedelta(days=interval)
+        elif recurring_rule_type == 'weekly':
+            return relativedelta(weeks=interval)
+        elif recurring_rule_type == 'monthly':
+            return relativedelta(months=interval)
+        elif recurring_rule_type == 'monthlylastday':
+            return relativedelta(months=interval, day=31)
+        else:
+            return relativedelta(years=interval)
+
+    @api.multi
+    def delay(self, delay_delta):
+        """
+        Delay a contract line
+        :param delay_delta: delay relative delta
+        :return: delayed contract line
+        """
+        for rec in self:
+            old_date_start = rec.date_start
+            old_date_end = rec.date_end
+            new_date_start = rec.date_start + delay_delta
+            rec.recurring_next_date = self._compute_first_recurring_next_date(
+                new_date_start,
+                rec.recurring_invoicing_type,
+                rec.recurring_rule_type,
+                rec.recurring_interval,
+            )
+            rec.date_end = (
+                rec.date_end
+                if not rec.date_end
+                else rec.date_end + delay_delta
+            )
+            rec.date_start = new_date_start
+            msg = _(
+                """Contract line for <strong>{product}</strong>
+                delayed: <br/>
+                - <strong>Start</strong>: {old_date_start} -- {new_date_start}
+                <br/>
+                - <strong>End</strong>: {old_date_end} -- {new_date_end}
+                """.format(
+                    product=rec.name,
+                    old_date_start=old_date_start,
+                    new_date_start=rec.date_start,
+                    old_date_end=old_date_end,
+                    new_date_end=rec.date_end,
+                )
+            )
+            rec.contract_id.message_post(body=msg)
+
+    @api.multi
+    def stop(self, date_end):
+        """
+        Put date_end on contract line
+        We don't consider contract lines that end's before the new end date
+        :param date_end: new date end for contract line
+        :return: True
+        """
+        if not all(self.mapped('is_stop_allowed')):
+            raise ValidationError(_('Stop not allowed for this line'))
+        for rec in self:
+            if date_end < rec.date_start:
+                rec.cancel()
+            else:
+                old_date_end = rec.date_end
+                date_end = (
+                    rec.date_end
+                    if rec.date_end and rec.date_end < date_end
+                    else date_end
+                )
```

Comment:
Maybe a question of taste but I think important business logic like this deserves an explicit if/then/else. Especially since we probably want to avoid write and message_post if the end date does not change.